server/manager/src/main/java/org/apache/accumulo/manager/Manager.java

```diff
@@ -568,10 +573,13 @@ private class MigrationCleanupThread implements Runnable {
     @Override
     public void run() {
       while (stillManager()) {
-        if (!migrations.isEmpty()) {
+        // migrations are stored in the metadata tables which cannot be read until the
+        // TabletGroupWatchers are started
+        if (watchersStarted.get() && numMigrations() > 0) {
           try {
             cleanupOfflineMigrations();
```

Comment:
It would probably be more efficient to collapse this function into cleanupDeletedMigrations() and scan the metadata table once.

```java
         for (var tabletMetadata : tabletsMetadata) {
                if(tabletMetadata.getMigration() != null && (!onlineTabletServers().contains(tabletMetadata.getMigration()) || OFFLINE = getTableState(tm.getExtent()))) {
                 tabletsMutator.mutateTablet(tabletMetadata.getExtent()).requireAbsentOperation().requireMigration(tabletMetadata.getMigration()).deleteMigration().submit(tm->false);
                }
              }
```